behave/runner.py

```diff
@@ -918,10 +918,21 @@ def load_hooks(self, filename=None):
     def load_step_definitions(self, extra_step_paths=None):
         if extra_step_paths is None:
             extra_step_paths = []
+
         # -- Allow steps to import other stuff from the steps dir
         # NOTE: Default matcher can be overridden in "environment.py" hook.
+        # Collect directories containing step files
         steps_dir = os.path.join(self.base_dir, self.config.steps_dir)
-        step_paths = [steps_dir] + list(extra_step_paths)
+        step_paths = set()  # Use a set to avoid duplicate paths
+
+        for root, _, files in os.walk(steps_dir):
+            if any(file.endswith('.py') and not file.startswith('__init__') for file in files):
+                step_paths.add(root)  # Add root directory containing .py step files
+
+        # Convert set to list and include extra_step_paths
+        step_paths = list(step_paths) + list(extra_step_paths)
```

Comment:
I like this code much better now! The last line is basically the same as the original line 924, the for-loop is straightforward and easy to understand. Technically, the `any()` hides the nested for-loop, but at least it's a pleasure to read.

On a side-comment, I don't think we don't need all those - pretty obvious - comments.

I'm still not sure, though, why we need to run the check over all files (to exclude directories that don't have Python source code). Also, is it a good idea to exclude `__init__.py` modules? Some people might want to put actual step code into it. -- Can you explain the underlying motivation?

Reply:
Originally, I included this check to ensure we only collect directories with valid Python step files, avoiding unnecessary or irrelevant ones. However, after considering your feedback, I agree that the check might not be essential, so I’ve removed it.

Reply:
* This approach leads to problems when the same `basename` of a Python module is used twice (as mentioned by @bittner above).
* The approach that you control the module naming makes it somewhat better, but does not solve this problem. When this constraint is violated by people that are unaware of this constraint, I wish you that you can resolve the problem in a few hours or days
* THEREFORE: This is not a valid solution as long as this problem is not solved.

Reply:
@jenisys In my last [commit](https://github.com/behave/behave/pull/1151/commits/8d85bc15243e617ae39f38552a3818e0a864d5f8) I believe I solved the issue. 
I also created a test for the scenario where we have a directory structure like this:
```
behave/
├── given
│   └── something.py
├── then
│   └── something.py
└── when
    └── something.py
```
This is as @bittner requested in this [comment](https://github.com/behave/behave/pull/1151#issuecomment-2438950824), and the tests have passed without any issues.